README.md
# Zenith

Perform gene set analysis on the result of differential expression using linear (mixed) modeling with [dream](https://doi.org/10.1093/bioinformatics/btaa687) by considering the correlation between gene expression traits.  This package implements the [camera](https://www.rdocumentation.org/packages/limma/versions/3.28.14/topics/camera) method from the [limma](https://bioconductor.org/packages/limma/) package proposed by [Wu and Smyth (2012)](https://doi.org/10.1093/nar/gks461).  `zenith()` is a simple extension of `camera()` to be compatible with linear mixed models implemented in `dream()`.


### Note
A future version will allow identification of gene sets with log fold changes with mixed sign. But this is not currently supported.

## Install

```r
library(devtools)

# install EnrichmentBrowser >= v2.19.13
install_github("lgeistlinger/EnrichmentBrowser")

# install variancePartition >= v1.19.16
install_github("GabrielHoffman/variancePartition")

install_github("GabrielHoffman/zenith")
```


### Issues
If you encounter the error:

```
Error in BroadCollection(category = tolower(cat), subCategory = tolower(subcat)) :
	invalid BroadCollection category: ‘c8’
```

This is due to a recent addition of [cell type signature gene sets](https://www.gsea-msigdb.org/gsea/msigdb/collections.jsp#C8) in [MSigDB](https://www.gsea-msigdb.org/gsea/msigdb) v7.2.  In order to be compatible with this new addition, you must update the R package `GSEABase`: 

```r
remotes::install_github("Bioconductor/GSEABase")
```
In the future, this will be installed by default


<sub><sub>Updated January 11, 2021</sub></sub>



NEWS
# Version 1.0.0
################
	o Improve documetation
	o fix bug with residuals

# Version 0.99.10
################
	o Upgrade to MSigDB 7.2

# Version 0.99.10
################
	o Update documentation

# Version 0.99.9
################
	o zenith report error when squaredStats is TRUE

# Version 0.99.8
################
	o Much faster loading of cached gene sets
	o enforce: 
	    - variancePartition (>= 1.19.16)
	    - EnrichmentBrowser (>= 2.19.13)

# Version 0.99.7
################
	o Enforce subsetting of residuals from MArrayLM

# Version 0.99.6
################
	o zenith() now returns effect size and standard error
	o if squaredStats==TRUE, return PValue = p.greater

# Version 0.99.5
################
	o Fix small error caused by fit$method

# Version 0.99.4
################
	o zenith uses t statistical for fixed effect models and z.std for mixed effects

# Version 0.99.3
################
	o zenith now works when dream uses random effects
	o more accurate time in progress bar

tests/runTests.R
BiocGenerics:::testPackage("zenith")

inst/unitTests/test_zenith.R
library(RUnit)

test_corInGeneSet = function(){

	library(variancePartition)

	set.seed(1)

	y <- matrix(rnorm(1000*6),1000,6)
	colnames(y) = paste0("sample_", 1:6)
	rownames(y) = paste0("gene_", 1:1000)

	design <- data.frame(Intercept=1,Group=c(0,0,0,1,1,1))
	rownames(design) = paste0("sample_", 1:6)

	# First set of 20 genes are genuinely differentially expressed
	index1 <- 1:20
	y[index1,4:6] <- y[index1,4:6]+1

	# Second set of 20 genes are not DE
	index2 <- 21:40

	# gene indeces
	lst = list(set1=index1,set2=index2)

	# differential expression with dream	
	fit = dream(y, ~ Group, design, computeResiduals=TRUE)
	fit = eBayes(fit)

	# Calculate correlation between residuals
	res1 = limma::interGeneCorrelation(y[index2,], design)

	res2 = zenith:::corInGeneSet(fit, rownames(fit)[index2])

	checkEquals(res1, res2)
}


test_zenith = function(){

	library(variancePartition)

	set.seed(1)

	y <- matrix(rnorm(1000*6),1000,6)
	colnames(y) = paste0("sample_", 1:6)
	rownames(y) = paste0("gene_", 1:1000)

	design <- data.frame(Intercept=1,Group=c(0,0,0,1,1,1))
	rownames(design) = paste0("sample_", 1:6)

	# First set of 20 genes are genuinely differentially expressed
	index1 <- 1:20
	y[index1,4:6] <- y[index1,4:6]+1

	# Second set of 20 genes are not DE
	index2 <- 21:40

	# gene indeces
	lst = list(set1=index1,set2=index2)

	# differential expression with dream	
	fit = dream(y, ~ Group, design, computeResiduals=TRUE)
	fit = eBayes(fit)

	# Parametric test
	res1 = camera(y, lst, design, inter.gene.cor=NA)
	res2 = zenith( fit, "Group", lst)

	# only check columns in common
	cols = colnames(res1)[1:5]

	test1 = checkEquals(res1[,cols], res2[,cols])

	# Non-parametric test
	res1 = camera(y, lst, design, inter.gene.cor=NA, use.ranks=TRUE)
	res2 = zenith( fit, "Group", lst, use.ranks=TRUE)

	test2 = checkEquals(res1[,cols], res2[,cols])

	test1 & test2
}




R/get_MSigDB.R

#' Load MSigDB genesets
#'
#' Load MSigDB genesets
#'
#' @param cat array of categories to load.  Defaults to array of all MSigDB categories
#' @param to convert gene names to this type using \code{EnrichmentBrowser::idMap()}.  See \code{EnrichmentBrowser::idTypes(org="hsa")} for valid types
#'
#' @details
#' This function loads the MSigDB gene sets using the packages \code{EnrichmentBrowser} and \code{msigdbr}.  It can take a mintute to load because converting gene name type is slow.   
#'
#' @return Gene sets stored as GeneSetCollection
#' @examples
#' # load Hallmark gene sets
#' gs = get_MSigDB('H')
#' 
#' # load all gene sets
#' # gs = get_MSigDB()
#' 
#' @import EnrichmentBrowser GSEABase msigdbr
#' @export
get_MSigDB = function(cat = unique(msigdbr_collections()$gs_cat), to = "ENSEMBL"){

	gs.list = lapply( cat, function(x){
		# get gene sets and convert to gene.id.type "to"
		getGenesets(org="hsa", db="msigdb", gene.id.type = to, return.type='GeneSetCollection', cat=x, subcat=NA)
	})

	# combine gene sets and convert to GeneSetCollection
	GeneSetCollection( do.call(c, gs.list ) )
}








R/corInGeneSet.R


#' Evaluate mean correlation between residuals in gene set
#'
#' Evaluate mean correlation between residuals in gene set based on results from dream
#'
#' @param fit result of differential expression with dream
#' @param idx indeces or rownames to extract
#' @param squareCorr compute the mean squared correlation instead
#'
#' @importFrom Rfast cora
corInGeneSet <- function( fit, idx, squareCorr = FALSE ){

  if( is.null(fit$residuals) ){
    stop("fit must be result of dream(..., computeResiduals=TRUE)")
  }

  # Extract residuals 
  resid = fit$residuals[idx,,drop=FALSE]

  m = nrow(resid)

  # evaluate correlation between residuals
  # This is the correlation between test statistics (almost, consider eBayes??)
  # C = cor(t(resid))
  C = cora(t(resid)) # this is a faster version of correlation in Rfast

  if( squareCorr ){
    # Correlation between squared test statistics
    # note that the covariance is 2*C^2
    C = C^2
  }

  # evaluate mean correlation
  correlation = mean(C[lower.tri(C)])

  vif = 1 + correlation*(m-1) 

  # if gene set has only 1 gene, correlation will be NAN
  # and VIF should be set to 1
  if( is.nan(correlation) ){
    correlation = NA
    vif = 1
  }

  list(vif=vif, correlation=correlation)
}






R/get_GeneOntology.R


#' Load Gene Ontology genesets
#'
#' Load Gene Ontology genesets
#'
#' @param onto array of categories to load
#' @param to convert gene names to this type using \code{EnrichmentBrowser::idMap()}.  See \code{EnrichmentBrowser::idTypes(org="hsa")} for valid types
#' @param includeOffspring if TRUE, follow the GO hierarchy down and include all genes in offspring sets for a given gene set
#'
#' @details
#' This function loads the GO gene sets using the packages \code{EnrichmentBrowser} and \code{GO.db}  It can take a mintute to load because converting gene name type is slow.   
#'
#' @return Gene sets stored as GeneSetCollection
#' @examples
#' # load GO Biological Process
#' gs = get_GeneOntology('BP')
#' 
#' # load all gene sets
#' # gs = get_GeneOntology()
#' 
#' @import EnrichmentBrowser GSEABase
#' @export
get_GeneOntology = function( onto = c("BP", "MF", "CC"), to = 'ENSEMBL', includeOffspring=TRUE ){

	gs.list = lapply( onto, function(x){
		getGenesets(org="hsa", db="go", gene.id.type = to, return.type='GeneSetCollection', onto=x)
		})

	# combine gene sets and convert to GeneSetCollection
	gs.go = GeneSetCollection( do.call(c,gs.list) )

	if( includeOffspring ){
		# follow the GO hierarchy down and include all genes in offspring sets for a given gene set
		gs.go = aggregate_GO_offspring( gs.go )
	}

	# summary(sapply(geneIds(gs.go), length))

	# modify gene set names to include description
	gs.go_rename = lapply( gs.go, function(gs){
	   setName(gs) = paste0( setName(gs), ': ', description(gs))
	   gs
	  } )
	GeneSetCollection( gs.go_rename )
}


#' Aggregate GO offspring genesets and genes
#'
#' Follow the GO hierarchy down and include all genes in offspring sets for a given gene set
#'
#' @param gs.GO GeneSetCollection from getGenesets()
#'
#' @return GeneSetCollection with aggregated gene sets
#'
#' @import EnrichmentBrowser GSEABase GO.db data.table
#'
aggregate_GO_offspring = function( gs.GO ){

	if( ! is(gs.GO, 'GeneSetCollection') ){
		stop("gsGO must be a GeneSetCollection")
	}

	# create a data.table of gene set names vs index for fast searching
	dt = data.table(GOID = gsub("GO", "GO:", names(gs.GO)), idx = seq_len(length(gs.GO)))
	setkey(dt, 'GOID')

	# summary(sapply(geneIds(gs.GO), length))

	# get combined of offspring for each gene set
	df_all = c(as.list(GOMFOFFSPRING), as.list(GOBPOFFSPRING), as.list(GOCCOFFSPRING))

	# for each gene set in GO, extract genes in this set and all offspring sets
	gs.GO.all_offspring = lapply(gs.GO, function(x){
		
		# get GO ID for this gene set
		goid = ids(collectionType(x))

		# get GO ID's or this and offspring gene sets
		query = c(goid, df_all[[goid]])
		query = query[!is.na(query)]

		# get the index through a fast look up in dt
		idx = dt[query,]$idx
		idx = idx[!is.na(idx)]

		gs.offspring = gs.GO[idx]

		# get gene array to current and offspring genes
		geneIds(x) = unique(unlist(geneIds(gs.offspring)))

		x
		})
	GeneSetCollection(gs.GO.all_offspring)
}






R/zenith.R


#' Gene set analysis following differential expression with dream
#'
#' Perform gene set analysis on the result of differential expression using linear (mixed) modeling with \code{variancePartition::dream} by considering the correlation between gene expression traits.  This package is a slight modification of \code{limma::camera} to 1) be compatible with dream, and 2) allow identification of gene sets with log fold changes with mixed sign.
#'
#' @param fit result of differential expression with dream
#' @param coef coefficient to test using \code{topTable(fit, coef)}
#' @param index an index vector or a list of index vectors.  Can be any vector such that \code{fit[index,]} selects the rows corresponding to the test set.  The list can be made using \code{ids2indices}.
#' @param use.ranks do a rank-based test (\code{TRUE}) or a parametric test ('FALSE')?
#' @param allow.neg.cor should reduced variance inflation factors be allowed for negative correlations?
# @param squaredStats Test squared test statstics to identify gene sets with log fold change of mixed sign.
#' @param progressbar if TRUE, show progress bar
#'
#' @details
#' \code{zenith} gives the same results as \code{camera(..., inter.gene.cor=NA)} which estimates the correlation with each gene set.  
#'
#' For differential expression with dream using linear (mixed) models see Hoffman and Roussos (2020).  For the original camera gene set test see Wu and Smyth (2012).
#' 
#' @references{
#'   \insertRef{hoffman2020dream}{zenith}
#' 
#'   \insertRef{wu2012camera}{zenith}
#' }
#'
#' @return
#' \itemize{
#'   \item \code{NGenes}: number of genes in this set
#'   \item \code{Correlation}: mean correlation between expression of genes in this set
#'   \item \code{delta}: difference in mean t-statistic for genes in this set compared to genes not in this set
#'   \item \code{se}: standard error of \code{delta}
#'   \item \code{p.less}: p-value for hypothesis test of \code{H0: delta < 0}
#'   \item \code{p.greater}: p-value for hypothesis test of \code{H0: delta > 0}
#'   \item \code{PValue}:  p-value for hypothesis test \code{H0: delta != 0}
#'   \item \code{Direction}: direction of effect based on sign(delta)
#'   \item \code{FDR}: false discovery rate based on Benjamini-Hochberg method in \code{p.adjust}
#' }
#'
#' @examples
#' 
#' library(variancePartition)
#' 
#' # simulate meta-data
#' info <- data.frame(Age=c(20, 31, 52, 35, 43, 45),Group=c(0,0,0,1,1,1))
#'
#' # simulate expression data
#' y <- matrix(rnorm(1000*6),1000,6)
#' rownames(y) = paste0("gene", 1:1000)
#' colnames(y) = rownames(info)
#' 
#' # First set of 20 genes are genuinely differentially expressed
#' index1 <- 1:20
#' y[index1,4:6] <- y[index1,4:6]+1
#' 
#' # Second set of 20 genes are not DE
#' index2 <- 21:40
#' 
#' # perform differential expression analysis with dream
#' fit = dream(y, ~ Age + (1|Group), info)
#' 
#' # perform gene set analysis testing Age
#' res = zenith(fit, "Age", list(set1=index1,set2=index2) )
#' 
#' head(res)
#' 
#' @importFrom Rdpack reprompt
#' @import limma variancePartition stats utils methods progress 
#'
#' @export
zenith <- function( fit, coef, index, use.ranks=FALSE, allow.neg.cor=FALSE, progressbar=TRUE ){

  if( ! is(fit, 'MArrayLM') ){
    stop("fit must be of class MArrayLM from variancePartition::dream")
  }

  if( is.null(fit$residuals) ){
    stop("fit must be result of dream(..., computeResiduals=TRUE)")
  }

  if( ! (coef %in% colnames(coef(fit))) ){
    stop("coef must be in colnames(coef(fit))")
  }

  if( is.null(rownames(fit)) ){
    stop("rownames(fit) is NULL.  Each feature must have a unique name")
  }

  # Check index
  if(!is.list(index)) index <- list(set1=index)
  nsets <- length(index)
  if(nsets==0L) stop("index is empty")

  # Only keep residuals for genes present in the main part of fit
  # Currently fit[1:10,] subsets objects in a standard MArrayLM
  #   but residuals is not standard, so it is not subsetted
  fit$residuals = fit$residuals[rownames(fit),,drop=FALSE]
  
  if( fit$method == "ls" ){

    # extract test statistics
    Stat = topTable(fit, coef, number=Inf, sort.by="none")$t

    if( ! use.ranks ){
      df = fit$df.total[1]
      Stat <- zscoreT( Stat, df=df, approx=TRUE, method="hill")
    }
  }else if( fit$method == "lmer"){
    # extract test statistics
    Stat = topTable(fit, coef, number=Inf, sort.by="none")$z.std
  }else{
    stop("Model method must be either 'ls' or 'lmer'")
  }

  # get number of statistics
  G = length( Stat )
  ID = rownames(fit)
  
  df.camera <- min(fit$df.residual[1], G - 2L)

  # Global statistics
  meanStat <- mean(Stat)
  varStat <- var(Stat)

  # setup progressbar
  if( progressbar ){
    # since time is quadratic in the size of the gene set
    total_work = sum(sapply(index, length)^2)

    pb <- progress_bar$new(
      format = " [:bar] :percent eta: :eta",
      clear = FALSE,
      total = total_work, width = 60)
  }
  cumulative_work = cumsum(sapply(index, length)^2)

  # tab <- matrix(0,nsets,5)
  # rownames(tab) <- names(index)
  # colnames(tab) <- c("NGenes","Correlation","Down","Up","TwoSided")
  tab = lapply(seq_len(nsets), function(i){    

    iset <- index[[i]]
    if(is.character(iset)) iset <- which(ID %in% iset)

    StatInSet <- Stat[iset]

    m <- length(StatInSet)
    m2 <- G-m

    # cumulative_work <<- cumulative_work + m^2

    # Compute correlation within geneset
    res = corInGeneSet( fit, iset, squareCorr=FALSE)
    correlation = res$correlation
    vif = res$vif

    # test
    # correlation = 0
    # vif = 1

    if(use.ranks) {

      corr.use = correlation
      if( ! allow.neg.cor ) corr.use <- max(0,corr.use)

      res = .rankSumTestWithCorrelation(iset, statistics=Stat, correlation=corr.use, df=df.camera)

      df = data.frame(  NGenes      = m,
                        Correlation = correlation,
                        delta       = res$effect,
                        se          = res$se,
                        p.less      = res$less,
                        p.greater   = res$greater )
    }else{ 

      if( ! allow.neg.cor ) vif <- max(1,vif)

      meanStatInSet <- mean(StatInSet)
      delta <- G/m2*(meanStatInSet-meanStat)
      varStatPooled <- ( (G-1)*varStat - delta^2*m*m2/G ) / (G-2)
      delta.se = sqrt( varStatPooled * (vif/m + 1/m2) )
      two.sample.t = delta / delta.se

      # two.sample.t <- delta / sqrt( varStatPooled * (vif/m + 1/m2) )
      # tab[i,3] <- pt(two.sample.t,df=df.camera)
      # tab[i,4] <- pt(two.sample.t,df=df.camera,lower.tail=FALSE)

      df = data.frame(NGenes      = m,
                      Correlation = correlation,
                      delta       = delta,
                      se          = delta.se,
                      p.less      = pt(two.sample.t,df=df.camera),
                      p.greater   = pt(two.sample.t,df=df.camera,lower.tail=FALSE) )
    }

    if( progressbar & (i %% 100 == 0) ) pb$update( cumulative_work[i] / total_work )

    df 
  })
  tab = do.call(rbind, tab)
  rownames(tab) <- names(index)

  # p-value for two sided test
  # if( squaredStats ){
  #   tab$PValue = tab$p.greater
  #   tab$Direction = "Up"
  # }else{    
    tab$PValue = 2*pmin(tab$p.less, tab$p.greater)
    tab$Direction = ifelse(tab$p.less < tab$p.greater, "Down", "Up")
  # }

  tab$FDR = p.adjust(tab$PValue, "BH")

  if( progressbar & ! pb$finished){
    pb$update( 1.0 )
    pb$terminate() 
  }

  # tab[,5] <- 2*pmin(tab[,3],tab[,4])

  # # New column names (Jan 2013)
  # tab <- data.frame(tab,stringsAsFactors=FALSE)
  # Direction <- rep_len("Up",length.out=nsets)
  # Direction[tab$Down < tab$Up] <- "Down"
  # tab$Direction <- Direction
  # tab$PValue <- tab$TwoSided
  # tab$Down <- tab$Up <- tab$TwoSided <- NULL

  # # Add FDR
  # if(nsets>1) tab$FDR <- p.adjust(tab$PValue,method="BH")

  # Sort by p-value
  o <- order(tab$PValue)
  tab <- tab[o,]

  tab
}




#' Two Sample Wilcoxon-Mann-Whitney Rank Sum Test Allowing For Correlation
#' 
#' Same as \code{limma::.rankSumTestWithCorrelation}, but returns effect size.
#'
#' @param index any index vector such that \code{statistics[index]} contains the values of the statistic for the test group.
#' @param statistics numeric vector giving values of the test statistic.
#' @param correlation numeric scalar, average correlation between cases in the test group.  Cases in the second group are assumed independent of each other and other the first group.
#' @param df degrees of freedom which the correlation has been estimated.
#'
#' @details See \code{limma::.rankSumTestWithCorrelation}
#'
.rankSumTestWithCorrelation = function (index, statistics, correlation = 0, df = Inf){
    n <- length(statistics)
    r <- rank(statistics)
    r1 <- r[index]
    n1 <- length(r1)
    n2 <- n - n1
    U <- n1 * n2 + n1 * (n1 + 1)/2 - sum(r1)
    mu <- n1 * n2/2
    if (correlation == 0 || n1 == 1) {
        sigma2 <- n1 * n2 * (n + 1)/12
    }
    else {
        sigma2 <- asin(1) * n1 * n2 + asin(0.5) * n1 * n2 * (n2 - 
            1) + asin(correlation/2) * n1 * (n1 - 1) * n2 * (n2 - 
            1) + asin((correlation + 1)/2) * n1 * (n1 - 1) * 
            n2
        sigma2 <- sigma2/2/pi
    }
    TIES <- (length(r) != length(unique(r)))
    if (TIES) {
        NTIES <- table(r)
        adjustment <- sum(NTIES * (NTIES + 1) * (NTIES - 1))/(n * 
            (n + 1) * (n - 1))
        sigma2 <- sigma2 * (1 - adjustment)
    }
    zlowertail <- (U + 0.5 - mu)/sqrt(sigma2)
    zuppertail <- (U - 0.5 - mu)/sqrt(sigma2)
    
    data.frame( effect  = -1*(U - mu), 
                se      = sqrt(sigma2),
                less    = pt(zuppertail, df = df, lower.tail = FALSE), 
                greater = pt(zlowertail, df = df))
}



















